Add post-processing classes

mikekryjak opened the pull request:
This PR adds two new classes, `FrontLocation` and `FrontLocationScan` providing useful post-processing of individual DLS results as well as an entire location scan. The DLS output is now slimmer and all the post-processing is done downstream. Some key variable names were also improved.

- [x] Merge in the new classes
- [x] Make them compatible
- [x] Update analytic_test and basic_use examples
- [x] Update geometry_manipulation and add test with single front position
- [x] Remove post-processing from the solver
- [x] Take the opportunity to rename some variables
- [x] Extended testing

Solves https://github.com/FusionDLS/FusionDLS/issues/63
Solves https://github.com/FusionDLS/FusionDLS/issues/56

mikekryjak committed:
Take out post-processing from solver

This is now done in FrontLocationScan

mikekryjak committed:
Add single front location test to examples

mikekryjak committed:
Tidy example

mikekryjak committed:
Replace trapz with trapezoid in postprocessing

mikekryjak committed:
Rename key output variables
New names are clearer and more consistent.

mikekryjak committed:
Add detachment window calcs to FrontLocationScan

mikekryjak committed:
Make control_variable a positional argument in SimulationInputs
The user should always be aware of and make a conscious decision for what they're solving for

mikekryjak committed:
Fix calculation of density profile

mikekryjak committed:
Include control_variable in inputs in the right place

mikekryjak committed:
Fix test_solver

control_variable is now a positional input

mikekryjak committed:
Save simulation inputs into postprocessing classes

mikekryjak committed:
Fix cvar in FrontLocation
Was previously reading from the state, which just captures the final state object for post-processing.... which means if you do several front positions, it'll have the cvar of the last position. I had some code written for the case of only one front position and tripped up on this.

LiamPattinson commented on src/fusiondls/postprocessing.py:
```diff
@@ -11,19 +11,26 @@ class FrontLocationScan:
 
-    def __init__(self, store):
-        num_locations = len(store["Sprofiles"])
+    def __init__(self, SimulationOutputs):
+        out = SimulationOutputs
```
I would reserve `CamelCase` for class/type names and use `snake_case` for variable names, with exceptions for things like physical quantities, where `t=time` and `T=temperature` may appear in the same equation (although in that case I'd lean towards just writing out `time` and `temperature` in full to remove all ambiguity). As this is very similarly named to `SimulationOutput`, it could be easily confused here for the class instead of the local variable. ([PEP8 recommendation](https://peps.python.org/pep-0008/#function-and-variable-names))

```suggestion
    def __init__(self, out: SimulationOutput):
```

mikekryjak replied:
Thanks, this is useful feedback. Fixed

LiamPattinson commented on src/fusiondls/postprocessing.py:
```diff
@@ -33,9 +40,11 @@ def __init__(self, store):
+            self.window = out["cvar"][-1] - out["cvar"][0]  # Cx - Ct
+            self.window_frac = (out["cvar"][-1] - out["cvar"][0]) / out["cvar"][
+                0
+            ]  # (Cx - Ct) / Ct
```
The formatter doesn't agree with the end-of-line comments here. I normally write comments on the line above to avoid these sorts of situations.

mikekryjak replied:
Done

LiamPattinson reviewed (approved):
Only have a couple of nit-picky comments for this one

mikekryjak committed:
Formatting

mikekryjak committed:
Fix postprocessing and improve docstrings

mikekryjak committed:
Increase rtol for basic_use test

It was failing on a relative error of 1e-5 for unknown reasons. An rtol of 1e-3 is more than sufficient.
